File: python-sdk/arc402/endpoint.py
"""HTTP endpoint helpers — resolve an agent's endpoint from AgentRegistry
and notify it after onchain events (hire, handshake).
"""
from __future__ import annotations

import json
import urllib.request
import urllib.error
from typing import Any

from web3 import Web3

from .abis import AgentRegistry_ABI

DEFAULT_REGISTRY_ADDRESS = "0xD5c2851B00090c92Ba7F4723FB548bb30C9B6865"


def resolve_endpoint(
    agent_address: str,
    rpc_url: str,
    registry_address: str = DEFAULT_REGISTRY_ADDRESS,
) -> str:
    """Read an agent's public HTTP endpoint from AgentRegistry.

    Returns an empty string if the agent is not registered or has no endpoint.
    """
    w3 = Web3(Web3.HTTPProvider(rpc_url))
    registry = w3.eth.contract(
        address=Web3.to_checksum_address(registry_address),
        abi=AgentRegistry_ABI,
    )
    agent_data = registry.functions.getAgent(
        Web3.to_checksum_address(agent_address)
    ).call()
    # agent_data is a tuple; endpoint is index 4
    if isinstance(agent_data, (list, tuple)):
        return str(agent_data[4]) if len(agent_data) > 4 else ""
    return str(getattr(agent_data, "endpoint", ""))


def notify_endpoint(
    endpoint: str,
    path: str,
    payload: dict[str, Any],
) -> dict[str, Any]:
    """POST a JSON payload to ``{endpoint}{path}``.

    Returns ``{"ok": True, "status": <int>}`` on success,
    ``{"ok": False, "error": <str>}`` on any failure. Never raises.
    """
    if not endpoint:
        return {"ok": False, "error": "no endpoint"}
    url = f"{endpoint}{path}"
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            return {"ok": True, "status": resp.status}
    except urllib.error.HTTPError as exc:
        return {"ok": False, "error": f"HTTP {exc.code}", "status": exc.code}
    except Exception as exc:  # noqa: BLE001
        return {"ok": False, "error": str(exc)}


def notify_hire(
    agent_address: str,
    proposal: dict[str, Any],
    rpc_url: str,
    registry_address: str = DEFAULT_REGISTRY_ADDRESS,
) -> dict[str, Any]:
    """Resolve agent endpoint and POST to /hire."""
    endpoint = resolve_endpoint(agent_address, rpc_url, registry_address)
    return notify_endpoint(endpoint, "/hire", proposal)


def notify_handshake(
    agent_address: str,
    payload: dict[str, Any],
    rpc_url: str,
    registry_address: str = DEFAULT_REGISTRY_ADDRESS,
) -> dict[str, Any]:
    """Resolve agent endpoint and POST to /handshake."""
    endpoint = resolve_endpoint(agent_address, rpc_url, registry_address)
    return notify_endpoint(endpoint, "/handshake", payload)

File: reference/sdk/src/endpoint.ts
/**
 * HTTP endpoint helpers — resolve an agent's endpoint from AgentRegistry
 * and notify it after onchain events (hire, handshake).
 */
import { ethers } from "ethers";
import { AGENT_REGISTRY_ABI } from "./contracts";

export const DEFAULT_REGISTRY_ADDRESS = "0xD5c2851B00090c92Ba7F4723FB548bb30C9B6865";

export interface EndpointNotifyResult {
  ok: boolean;
  status?: number;
  error?: string;
}

/**
 * Reads an agent's public HTTP endpoint from AgentRegistry.
 * Returns an empty string if the agent is not registered or has no endpoint.
 */
export async function resolveEndpoint(
  agentAddress: string,
  provider: ethers.Provider,
  registryAddress = DEFAULT_REGISTRY_ADDRESS
): Promise<string> {
  const registry = new ethers.Contract(registryAddress, AGENT_REGISTRY_ABI, provider);
  const agentData = await registry.getAgent(agentAddress);
  return (agentData.endpoint as string) ?? "";
}

/**
 * POSTs a JSON payload to `${endpoint}${path}`.
 * Returns { ok, status } on success, { ok: false, error } on failure.
 * Never throws.
 */
export async function notifyEndpoint(
  endpoint: string,
  path: string,
  payload: Record<string, unknown>
): Promise<EndpointNotifyResult> {
  if (!endpoint) return { ok: false, error: "no endpoint" };
  try {
    const res = await fetch(`${endpoint}${path}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    return { ok: res.ok, status: res.status };
  } catch (err) {
    return { ok: false, error: err instanceof Error ? err.message : String(err) };
  }
}

/**
 * Convenience: resolve an agent's endpoint then POST to /hire.
 */
export async function notifyHire(
  agentAddress: string,
  proposal: Record<string, unknown>,
  provider: ethers.Provider,
  registryAddress = DEFAULT_REGISTRY_ADDRESS
): Promise<EndpointNotifyResult> {
  const endpoint = await resolveEndpoint(agentAddress, provider, registryAddress);
  return notifyEndpoint(endpoint, "/hire", proposal);
}

/**
 * Convenience: resolve an agent's endpoint then POST to /handshake.
 */
export async function notifyHandshake(
  agentAddress: string,
  payload: Record<string, unknown>,
  provider: ethers.Provider,
  registryAddress = DEFAULT_REGISTRY_ADDRESS
): Promise<EndpointNotifyResult> {
  const endpoint = await resolveEndpoint(agentAddress, provider, registryAddress);
  return notifyEndpoint(endpoint, "/handshake", payload);
}

/**
 * Convenience: resolve an agent's endpoint then POST to /hire/accepted.
 */
export async function notifyHireAccepted(
  agentAddress: string,
  payload: Record<string, unknown>,
  provider: ethers.Provider,
  registryAddress = DEFAULT_REGISTRY_ADDRESS
): Promise<EndpointNotifyResult> {
  const endpoint = await resolveEndpoint(agentAddress, provider, registryAddress);
  return notifyEndpoint(endpoint, "/hire/accepted", payload);
}

/**
 * Convenience: resolve an agent's endpoint then POST to /delivery.
 */
export async function notifyDelivery(
  agentAddress: string,
  payload: Record<string, unknown>,
  provider: ethers.Provider,
  registryAddress = DEFAULT_REGISTRY_ADDRESS
): Promise<EndpointNotifyResult> {
  const endpoint = await resolveEndpoint(agentAddress, provider, registryAddress);
  return notifyEndpoint(endpoint, "/delivery", payload);
}

/**
 * Convenience: resolve an agent's endpoint then POST to /delivery/accepted.
 */
export async function notifyDeliveryAccepted(
  agentAddress: string,
  payload: Record<string, unknown>,
  provider: ethers.Provider,
  registryAddress = DEFAULT_REGISTRY_ADDRESS
): Promise<EndpointNotifyResult> {
  const endpoint = await resolveEndpoint(agentAddress, provider, registryAddress);
  return notifyEndpoint(endpoint, "/delivery/accepted", payload);
}

/**
 * Convenience: resolve an agent's endpoint then POST to /dispute.
 */
export async function notifyDispute(
  agentAddress: string,
  payload: Record<string, unknown>,
  provider: ethers.Provider,
  registryAddress = DEFAULT_REGISTRY_ADDRESS
): Promise<EndpointNotifyResult> {
  const endpoint = await resolveEndpoint(agentAddress, provider, registryAddress);
  return notifyEndpoint(endpoint, "/dispute", payload);
}

/**
 * Convenience: resolve an agent's endpoint then POST to /message.
 */
export async function notifyMessage(
  agentAddress: string,
  payload: Record<string, unknown>,
  provider: ethers.Provider,
  registryAddress = DEFAULT_REGISTRY_ADDRESS
): Promise<EndpointNotifyResult> {
  const endpoint = await resolveEndpoint(agentAddress, provider, registryAddress);
  return notifyEndpoint(endpoint, "/message", payload);
}
